### ISSUE TITLE
Users are hitting Gradle failures when upgrading Flutter to v1.12.13+hotfix.5

### ISSUE THREAD

**Issue by amirh** [flutter/flutter#46898]
There are multiple reports of users hitting Gradle failures following the upgrade (#46807, #46753, #46790).

The root cause is that we've added platform specific plugin packages for web and macos with no `android/` folders as dependencies of popular plugins (url_launcher, google_sign_in, video_player, firebase_auth, and firebase_core).

Flutter's Android build configuration used to assume that every plugin has a buildable `android/` folder. We have landed https://github.com/flutter/flutter/pull/40640 https://github.com/flutter/flutter/pull/46282 to mitigate this problem however there are still 2 classes of issues:

## Cached build artifact
It's not clear to me exactly what, but it seems like right after upgrading flutter, the build system still tries to look for the buildable `android/` folder (e.g #46282 does not take immediate effect), and users have to delete some locally cached state before it takes affect. It's unclear to me what cached state exactly matters, some users report that deleting `pubspec.lock` and running `flutter clean` solved the issue, and some report that they had to clean their `.pub-cache` folder.

This is a nuisance for many users who upgrade flutter and try to build their existing app. We should better understand which cached state is causing the problem and figure out how to mitigate. @blasten @jonahwilliams do you have any insights?

## The aar build strategy
Flutter has a fallback build strategy to mitigate AndroidX issues, if the Android build fails Flutter may try a secondary build strategy which involves building each plugin as AAR, and then building the app. This aar build strategy was not updated to support non-Android plugins, and is failing when the project has a non-Android plugin dependency.

Note that this scenario is expected to be less common, it will fail for applications where the initial build fails due to an AndroidX failure and also have a non-Android plugin dependency.

@blasten has sent out https://github.com/flutter/flutter/pull/46835 to mitigate that issue, however it's yet clear if that fix qualifies for a hotfix.

## Temporary mitigation in plugins code
One temporary mitigation which should address both issues would be to add a no-op `android/` folder to all plugins, until there's a new stable release that addresses both issues. I suggest we start adding these to packages of the plugins mentioned above as a temporary remedy. Once the issues above are solved on stable we can clean these up.

**Comment by mklim** [comment]
@amirh just to be clear so I understand what needs to be mitigated immediately, these are all the plugins affected, correct?

- [ ] url_launcher
- [ ] google_sign_in
- [ ] video_player
- [ ] firebase_auth
- [ ] firebase_core

**Comment by jonahwilliams** [comment]
> , but it seems like right after upgrade flutter, the build system still tried to look for the buildable android/ folder (e.g #46282 does not take immediate effect), and users have to delete some locally cached state before it takes affect.

Possible the gradle.daemon is still running the old gradle script. If we can reproduce it locally, we could test by manually killing the daemon and re-running

**Comment by amirh** [comment]
@mklim I updated the plug

> @amirh just to be clear so I understand what needs to be mitigated immediately, these are all the plugins affected, correct?
> 

Correct, I updated the list with the specific packages that need to be updated

**Comment by mklim** [comment]
@amirh Thanks. I think it would be a good idea to create sub issues for mitigating specific plugins so we can track who's fixing which one. I'll start with google_sign_in_web next.

**Comment by feinstein** [comment]
I am having this issue too, may I suggest a port-mortem analysis on this? 

I think the Flutter team is incredibly capable and experienced, but I can't help myself on thinking how could a major release be shipped without anyone seeing any impact on so many popular packages. Maybe a new test protocol for releases should be studied?

I don't intent on criticize anyone, I love your work from the bottom of my heart, I just like to take opportunities like this to improve and come out with a stronger and more robust product.

**Comment by amirh** [comment]
Good idea, I'll follow up with a post-mortem.

**Comment by mklim** [comment]
Important caveat for people that are working on the mitigations (@bparrishMines @ditman):

Like always, it's possible to introduce AndroidX failures in even the boilerplate plugin code. We should manually update any patches to avoid any references to `androidx` or `support` entirely in the non-op Android code. In this case the code should never be executed so we don't need to use either.

As a starting point you can create the Android template pretty easily with `flutter create` and then remove some `support` references in the source code from there.

```terminal
$ cd /tmp
$ flutter create -t plugin -i objc -a java --no-androidx --org io.flutter.plugins  $plugin
$ cp -r $plugin/android path/to/$plugin
```

**Comment by lacu** [comment]
I am going crazy!!!, I have a broken app. 
It's possible downgrade the sdk, I cloned 1.12.13.hotfix.3 from git, but flutter doctor, made the upgrade again!!!

**Comment by amirh** [comment]
@lacu I'm sorry, we're working to re-publish all affected plugins ASAP, will hopefully be done within a few hours (mainly waiting for CI right now). If you can't wait a few hours you should be able to temporarily pin the offending plugin's version (assuming you're not building for web or macos) to the version before the web/macos endorsement, don't forget to unpin once the workarounds are published.

**Comment by blasten** [comment]
> This is a nuisance for many users who upgrade flutter and try to build their existing app. We should better understand which cached state is causing the problem and figure out how to mitigate.

@amirh is the case that a user was on dev (without #46282) and then switched to stable? or does this happen after a stable -> stable upgrade?

**Comment by amirh** [comment]
@blasten I don't know, asked on https://github.com/flutter/flutter/issues/46807#issuecomment-565181760

[from the volume of thumbs I'd imagine it's stable -> stable, but I may very well be wrong]

**Comment by amirh** [comment]
Status update: all `flutter/plugins` plugins listed above have been published with the workaround. firebase_core and firebase_auth are still waiting for CI.

**Comment by amirh** [comment]
https://github.com/flutter/flutter/issues/46807#issuecomment-565196488 - confirms that  stable->stable upgrades were affected

**Comment by lacu** [comment]
there is problems with flutter_map too? , I having this errors whe update to 1.12.13 hotfix 5

Error: The method 'NetworkImageWithRetry.load' has fewer positional arguments than those of overridden method 'ImageProvider.load'.
  ImageStreamCompleter load(NetworkImageWithRetry key) {

**Comment by amirh** [comment]
@lacu that seems unrelated, see the note at the top of the [readme for cached_network_image](https://pub.dev/packages/cached_network_image), seems like they recommend upgrading to 2.0.0-rc

**Comment by cyanglaz** [comment]
The fix on firebase_core_web and firebase_auth_web are published. Please give it a try see if the issues are fixed.

**Comment by lacu** [comment]
after hours, the solution was downgrade to 1.9.1

**Comment by workerbee22** [comment]
Also have this issue with upgrade Stable 1.9.1 -> 1.12.13+hotfix.5

**Comment by feinstein** [comment]
I don't know if this is related, but I was having crashes, so I updated to the latest crashlytics and auth versions, then I got more crashes, but with different messages, but the code was compiling and executing, even with the crash messages... Then after some builds the crashes just vanished.... I don't remember if I did something, I think they just changed to "using a deprecated API" message but were shown only once and now I don't get them.

But I am getting all my previous exceptions logged by crashlytics being dump in the console when I am building my app. It stops for a while and later it happens again, I don't know what's going on.

**Comment by amirh** [comment]
With the latest versions of the plugins mentioned above users should not be hitting these errors any more.

If you're still getting the errors please try to upgrade your packages by running `flutter pub update` in the application folder.

I'm removing the `customer: blocker` label but keeping the issue open to follow-up with a postmortem.

Please leave a comment if after upgrading you're still hitting Gradle errors like:
```
Could not determine the dependencies of task ':firebase_core:compileDebugAidl'.
> Could not resolve all task dependencies for configuration ':firebase_core:debugCompileClasspath'.
   > Could not resolve project :firebase_core_web.
     Required by:
         project :firebase_core
      > Unable to find a matching configuration of project :firebase_core_web:
          - None of the consumable configurations have attributes.
```

**Comment by feinstein** [comment]
@amirh could you take a look if #46985 is related to this?

**Comment by lacu** [comment]
I lost a lot hours trying to fix the problems. 
I made a downgrade to 1.9.1, and all run again.

**Comment by amirh** [comment]
> @amirh could you take a look if #46985 is related to this?

Shouldn't be related (the current issue leads to Gradle build failures, #46985 doesn't seem to describe a Gradle failure).

**Comment by feinstein** [comment]
Could you please forward it to anyone that could be related to this? I am not sure if its related to crashlytics, or the `flutter` tool or other part of the framework.

**Comment by amirh** [comment]
@feinstein let's keep this thread on-topic. If #46985 is a flutter bug that issue is the right place to discuss it. I'd suggest including a minimal code snippet that reproduces that other issue.

**Comment by akarabach** [comment]
> With the latest versions of the plugins mentioned above users should not be hitting these errors any more.
> 
> If you're still getting the errors please try to upgrade your packages by running `flutter pub update` in the application folder.
> 
> I'm removing the `customer: blocker` label but keeping the issue open to follow-up with a postmortem.
> 
> Please leave a comment if after upgrading you're still hitting Gradle errors like:
> 
> ```
> Could not determine the dependencies of task ':firebase_core:compileDebugAidl'.
> > Could not resolve all task dependencies for configuration ':firebase_core:debugCompileClasspath'.
>    > Could not resolve project :firebase_core_web.
>      Required by:
>          project :firebase_core
>       > Unable to find a matching configuration of project :firebase_core_web:
>           - None of the consumable configurations have attributes.
> ```

flutter pub update - doesn't exist ( error for console `Could not find a subcommand named "update" for "flutter pub".`)

**Comment by feinstein** [comment]
In IntelliJ IDEA or Android Studio, go to `Tools`>`Flutter`>`Flutter Packages Upgrade`

**Comment by akarabach** [comment]
I still receive errors, I cleaned my cache manually, I runed `flutter clean` and `flutter pub cache repair`, I upgraded package but it not helps.
`Launching lib\main.dart on Android SDK built for x86 in debug mode...
Running Gradle task 'assembleDebug'...
I/flutter ( 2455): AsyncSnapshot<List<dynamic>>(ConnectionState.waiting, [], null)
I/flutter ( 2455): AsyncSnapshot<List<dynamic>>(ConnectionState.active, [], null)
I/flutter ( 2455): AsyncSnapshot<List<dynamic>>(ConnectionState.waiting, [], null)
I/flutter ( 2455): {firstName: qwe, lastName: qwe, username: qwe, timestamp: Timestamp(seconds=1575231591, nanoseconds=478000000), photoUrl: photourl}
I/flutter ( 2455): AsyncSnapshot<List<dynamic>>(ConnectionState.waiting, [], null)
I/flutter ( 2455): AsyncSnapshot<List<dynamic>>(ConnectionState.active, [], null)
I/flutter ( 2455): AsyncSnapshot<List<dynamic>>(ConnectionState.waiting, [], null)
I/flutter ( 2455): AsyncSnapshot<List<dynamic>>(ConnectionState.waiting, [], null)
I/flutter ( 2455): AsyncSnapshot<List<dynamic>>(ConnectionState.active, [], null)
E/flutter ( 6230): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: ServicesBinding.defaultBinaryMessenger was accessed before the binding was initialized.
E/flutter ( 6230): If you're running an application and need to access the binary messenger before `runApp()` has been called (for example, during plugin initialization), then you need to explicitly call the `WidgetsFlutterBinding.ensureInitialized()` first.
E/flutter ( 6230): If you're running a test, you can call the `TestWidgetsFlutterBinding.ensureInitialized()` as the first line in your test's `main()` method to initialize the binding.
E/flutter ( 6230): #0      defaultBinaryMessenger.<anonymous closure> (package:flutter/src/services/binary_messenger.dart:76:7)
E/flutter ( 6230): #1      defaultBinaryMessenger (package:flutter/src/services/binary_messenger.dart:89:4)
E/flutter ( 6230): #2      MethodChannel.binaryMessenger (package:flutter/src/services/platform_channel.dart:140:62)
E/flutter ( 6230): #3      MethodChannel.setMethodCallHandler (package:flutter/src/services/platform_channel.dart:368:5)
E/flutter ( 6230): #4      new Firestore (package:cloud_firestore/src/firestore.dart:13:13)
E/flutter ( 6230): #5      Firestore.instance (package:cloud_firestore/src/firestore.dart:39:37)
E/flutter ( 6230): #6      Firestore.instance (package:cloud_firestore/src/firestore.dart:39:26)
E/flutter ( 6230): #7      new UserRepository (package:example_application/repository/user.dart:11:49)
E/flutter ( 6230): #8      new ExampleApplicationApp (package:example_application/main.dart:33:42)
E/flutter ( 6230): #9      main (package:example_application/main.dart:29:10)
E/flutter ( 6230): #10     _runMainZoned.<anonymous closure>.<anonymous closure> (dart:ui/hooks.dart:239:25)
E/flutter ( 6230): #11     _rootRun (dart:async/zone.dart:1126:13)
E/flutter ( 6230): #12     _CustomZone.run (dart:async/zone.dart:1023:19)
E/flutter ( 6230): #13     _runZoned (dart:async/zone.dart:1518:10)
E/flutter ( 6230): #14     runZoned (dart:async/zone.dart:1502:12)
E/flutter ( 6230): #15     _runMainZoned.<anonymous closure> (dart:ui/hooks.dart:231:5)
E/flutter ( 6230): #16     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:307:19)
E/flutter ( 6230): #17     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:174:12)
E/flutter ( 6230): 
E/flutter ( 6407): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: ServicesBinding.defaultBinaryMessenger was accessed before the binding was initialized.
E/flutter ( 6407): If you're running an application and need to access the binary messenger before `runApp()` has been called (for example, during plugin initialization), then you need to explicitly call the `WidgetsFlutterBinding.ensureInitialized()` first.
E/flutter ( 6407): If you're running a test, you can call the `TestWidgetsFlutterBinding.ensureInitialized()` as the first line in your test's `main()` method to initialize the binding.
E/flutter ( 6407): #0      defaultBinaryMessenger.<anonymous closure> (package:flutter/src/services/binary_messenger.dart:76:7)
E/flutter ( 6407): #1      defaultBinaryMessenger (package:flutter/src/services/binary_messenger.dart:89:4)
E/flutter ( 6407): #2      MethodChannel.binaryMessenger (package:flutter/src/services/platform_channel.dart:140:62)
E/flutter ( 6407): #3      MethodChannel.setMethodCallHandler (package:flutter/src/services/platform_channel.dart:368:5)
E/flutter ( 6407): #4      new Firestore (package:cloud_firestore/src/firestore.dart:13:13)
E/flutter ( 6407): #5      Firestore.instance (package:cloud_firestore/src/firestore.dart:39:37)
E/flutter ( 6407): #6      Firestore.instance (package:cloud_firestore/src/firestore.dart:39:26)
E/flutter ( 6407): #7      new UserRepository (package:example_application/repository/user.dart:11:49)
E/flutter ( 6407): #8      new ExampleApplicationApp (package:example_application/main.dart:33:42)
E/flutter ( 6407): #9      main (package:example_application/main.dart:29:10)
E/flutter ( 6407): #10     _runMainZoned.<anonymous closure>.<anonymous closure> (dart:ui/hooks.dart:239:25)
E/flutter ( 6407): #11     _rootRun (dart:async/zone.dart:1126:13)
E/flutter ( 6407): #12     _CustomZone.run (dart:async/zone.dart:1023:19)
E/flutter ( 6407): #13     _runZoned (dart:async/zone.dart:1518:10)
E/flutter ( 6407): #14     runZoned (dart:async/zone.dart:1502:12)
E/flutter ( 6407): #15     _runMainZoned.<anonymous closure> (dart:ui/hooks.dart:231:5)
E/flutter ( 6407): #16     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:307:19)
E/flutter ( 6407): #17     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:174:12)
E/flutter ( 6407): 
E/flutter ( 6535): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: ServicesBinding.defaultBinaryMessenger was accessed before the binding was initialized.
E/flutter ( 6535): If you're running an application and need to access the binary messenger before `runApp()` has been called (for example, during plugin initialization), then you need to explicitly call the `WidgetsFlutterBinding.ensureInitialized()` first.
E/flutter ( 6535): If you're running a test, you can call the `TestWidgetsFlutterBinding.ensureInitialized()` as the first line in your test's `main()` method to initialize the binding.
E/flutter ( 6535): #0      defaultBinaryMessenger.<anonymous closure> (package:flutter/src/services/binary_messenger.dart:76:7)
E/flutter ( 6535): #1      defaultBinaryMessenger (package:flutter/src/services/binary_messenger.dart:89:4)
E/flutter ( 6535): #2      MethodChannel.binaryMessenger (package:flutter/src/services/platform_channel.dart:140:62)
E/flutter ( 6535): #3      MethodChannel.setMethodCallHandler (package:flutter/src/services/platform_channel.dart:368:5)
E/flutter ( 6535): #4      new Firestore (package:cloud_firestore/src/firestore.dart:13:13)
E/flutter ( 6535): #5      Firestore.instance (package:cloud_firestore/src/firestore.dart:39:37)
E/flutter ( 6535): #6      Firestore.instance (package:cloud_firestore/src/firestore.dart:39:26)
E/flutter ( 6535): #7      new UserRepository (package:example_application/repository/user.dart:11:49)
E/flutter ( 6535): #8      new ExampleApplicationApp (package:example_application/main.dart:33:42)
E/flutter ( 6535): #9      main (package:example_application/main.dart:29:10)
E/flutter ( 6535): #10     _runMainZoned.<anonymous closure>.<anonymous closure> (dart:ui/hooks.dart:239:25)
E/flutter ( 6535): #11     _rootRun (dart:async/zone.dart:1126:13)
E/flutter ( 6535): #12     _CustomZone.run (dart:async/zone.dart:1023:19)
E/flutter ( 6535): #13     _runZoned (dart:async/zone.dart:1518:10)
E/flutter ( 6535): #14     runZoned (dart:async/zone.dart:1502:12)
E/flutter ( 6535): #15     _runMainZoned.<anonymous closure> (dart:ui/hooks.dart:231:5)
E/flutter ( 6535): #16     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:307:19)
E/flutter ( 6535): #17     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:174:12)
E/flutter ( 6535): 
E/flutter ( 6654): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: ServicesBinding.defaultBinaryMessenger was accessed before the binding was initialized.
E/flutter ( 6654): If you're running an application and need to access the binary messenger before `runApp()` has been called (for example, during plugin initialization), then you need to explicitly call the `WidgetsFlutterBinding.ensureInitialized()` first.
E/flutter ( 6654): If you're running a test, you can call the `TestWidgetsFlutterBinding.ensureInitialized()` as the first line in your test's `main()` method to initialize the binding.
E/flutter ( 6654): #0      defaultBinaryMessenger.<anonymous closure> (package:flutter/src/services/binary_messenger.dart:76:7)
E/flutter ( 6654): #1      defaultBinaryMessenger (package:flutter/src/services/binary_messenger.dart:89:4)
E/flutter ( 6654): #2      MethodChannel.binaryMessenger (package:flutter/src/services/platform_channel.dart:140:62)
E/flutter ( 6654): #3      MethodChannel.setMethodCallHandler (package:flutter/src/services/platform_channel.dart:368:5)
E/flutter ( 6654): #4      new Firestore (package:cloud_firestore/src/firestore.dart:13:13)
E/flutter ( 6654): #5      Firestore.instance (package:cloud_firestore/src/firestore.dart:39:37)
E/flutter ( 6654): #6      Firestore.instance (package:cloud_firestore/src/firestore.dart:39:26)
E/flutter ( 6654): #7      new UserRepository (package:example_application/repository/user.dart:11:49)
E/flutter ( 6654): #8      new ExampleApplicationApp (package:example_application/main.dart:33:42)
E/flutter ( 6654): #9      main (package:example_application/main.dart:29:10)
E/flutter ( 6654): #10     _runMainZoned.<anonymous closure>.<anonymous closure> (dart:ui/hooks.dart:239:25)
E/flutter ( 6654): #11     _rootRun (dart:async/zone.dart:1126:13)
E/flutter ( 6654): #12     _CustomZone.run (dart:async/zone.dart:1023:19)
E/flutter ( 6654): #13     _runZoned (dart:async/zone.dart:1518:10)
E/flutter ( 6654): #14     runZoned (dart:async/zone.dart:1502:12)
E/flutter ( 6654): #15     _runMainZoned.<anonymous closure> (dart:ui/hooks.dart:231:5)
E/flutter ( 6654): #16     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:307:19)
E/flutter ( 6654): #17     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:174:12)
E/flutter ( 6654): 
E/flutter ( 6654): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: ServicesBinding.defaultBinaryMessenger was accessed before the binding was initialized.
E/flutter ( 6654): If you're running an application and need to access the binary messenger before `runApp()` has been called (for example, during plugin initialization), then you need to explicitly call the `WidgetsFlutterBinding.ensureInitialized()` first.
E/flutter ( 6654): If you're running a test, you can call the `TestWidgetsFlutterBinding.ensureInitialized()` as the first line in your test's `main()` method to initialize the binding.
E/flutter ( 6654): #0      defaultBinaryMessenger.<anonymous closure> (package:flutter/src/services/binary_messenger.dart:76:7)
E/flutter ( 6654): #1      defaultBinaryMessenger (package:flutter/src/services/binary_messenger.dart:89:4)
E/flutter ( 6654): #2      MethodChannel.binaryMessenger (package:flutter/src/services/platform_channel.dart:140:62)
E/flutter ( 6654): #3      MethodChannel.setMethodCallHandler (package:flutter/src/services/platform_channel.dart:368:5)
E/flutter ( 6654): #4      new Firestore (package:cloud_firestore/src/firestore.dart:13:13)
E/flutter ( 6654): #5      Firestore.instance (package:cloud_firestore/src/firestore.dart:39:37)
E/flutter ( 6654): #6      Firestore.instance (package:cloud_firestore/src/firestore.dart:39:26)
E/flutter ( 6654): #7      new UserRepository (package:example_application/repository/user.dart:11:49)
E/flutter ( 6654): #8      new ExampleApplicationApp (package:example_application/main.dart:33:42)
E/flutter ( 6654): #9      main (package:example_application/main.dart:29:10)
E/flutter ( 6654): #10     _runMainZoned.<anonymous closure>.<anonymous closure> (dart:ui/hooks.dart:239:25)
E/flutter ( 6654): #11     _rootRun (dart:async/zone.dart:1126:13)
E/flutter ( 6654): #12     _CustomZone.run (dart:async/zone.dart:1023:19)
E/flutter ( 6654): #13     _runZoned (dart:async/zone.dart:1518:10)
E/flutter ( 6654): #14     runZoned (dart:async/zone.dart:1502:12)
E/flutter ( 6654): #15     _runMainZoned.<anonymous closure> (dart:ui/hooks.dart:231:5)
E/flutter ( 6654): #16     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:307:19)
E/flutter ( 6654): #17     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:174:12)
E/flutter ( 6654): 
E/flutter ( 6840): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: ServicesBinding.defaultBinaryMessenger was accessed before the binding was initialized.
E/flutter ( 6840): If you're running an application and need to access the binary messenger before `runApp()` has been called (for example, during plugin initialization), then you need to explicitly call the `WidgetsFlutterBinding.ensureInitialized()` first.
E/flutter ( 6840): If you're running a test, you can call the `TestWidgetsFlutterBinding.ensureInitialized()` as the first line in your test's `main()` method to initialize the binding.
E/flutter ( 6840): #0      defaultBinaryMessenger.<anonymous closure> (package:flutter/src/services/binary_messenger.dart:76:7)
E/flutter ( 6840): #1      defaultBinaryMessenger (package:flutter/src/services/binary_messenger.dart:89:4)
E/flutter ( 6840): #2      MethodChannel.binaryMessenger (package:flutter/src/services/platform_channel.dart:140:62)
E/flutter ( 6840): #3      MethodChannel.setMethodCallHandler (package:flutter/src/services/platform_channel.dart:368:5)
E/flutter ( 6840): #4      new Firestore (package:cloud_firestore/src/firestore.dart:13:13)
E/flutter ( 6840): #5      Firestore.instance (package:cloud_firestore/src/firestore.dart:39:37)
E/flutter ( 6840): #6      Firestore.instance (package:cloud_firestore/src/firestore.dart:39:26)
E/flutter ( 6840): #7      new UserRepository (package:example_application/repository/user.dart:11:49)
E/flutter ( 6840): #8      new ExampleApplicationApp (package:example_application/main.dart:33:42)
E/flutter ( 6840): #9      main (package:example_application/main.dart:29:10)
E/flutter ( 6840): #10     _runMainZoned.<anonymous closure>.<anonymous closure> (dart:ui/hooks.dart:239:25)
E/flutter ( 6840): #11     _rootRun (dart:async/zone.dart:1126:13)
E/flutter ( 6840): #12     _CustomZone.run (dart:async/zone.dart:1023:19)
E/flutter ( 6840): #13     _runZoned (dart:async/zone.dart:1518:10)
E/flutter ( 6840): #14     runZoned (dart:async/zone.dart:1502:12)
E/flutter ( 6840): #15     _runMainZoned.<anonymous closure> (dart:ui/hooks.dart:231:5)
E/flutter ( 6840): #16     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:307:19)
E/flutter ( 6840): #17     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:174:12)
E/flutter ( 6840): 
E/flutter ( 7100): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: ServicesBinding.defaultBinaryMessenger was accessed before the binding was initialized.
E/flutter ( 7100): If you're running an application and need to access the binary messenger before `runApp()` has been called (for example, during plugin initialization), then you need to explicitly call the `WidgetsFlutterBinding.ensureInitialized()` first.
E/flutter ( 7100): If you're running a test, you can call the `TestWidgetsFlutterBinding.ensureInitialized()` as the first line in your test's `main()` method to initialize the binding.
E/flutter ( 7100): #0      defaultBinaryMessenger.<anonymous closure> (package:flutter/src/services/binary_messenger.dart:76:7)
E/flutter ( 7100): #1      defaultBinaryMessenger (package:flutter/src/services/binary_messenger.dart:89:4)
E/flutter ( 7100): #2      MethodChannel.binaryMessenger (package:flutter/src/services/platform_channel.dart:140:62)
E/flutter ( 7100): #3      MethodChannel.setMethodCallHandler (package:flutter/src/services/platform_channel.dart:368:5)
E/flutter ( 7100): #4      new Firestore (package:cloud_firestore/src/firestore.dart:13:13)
E/flutter ( 7100): #5      Firestore.instance (package:cloud_firestore/src/firestore.dart:39:37)
E/flutter ( 7100): #6      Firestore.instance (package:cloud_firestore/src/firestore.dart:39:26)
E/flutter ( 7100): #7      new UserRepository (package:example_application/repository/user.dart:11:49)
E/flutter ( 7100): #8      new ExampleApplicationApp (package:example_application/main.dart:33:42)
E/flutter ( 7100): #9      main (package:example_application/main.dart:29:10)
E/flutter ( 7100): #10     _runMainZoned.<anonymous closure>.<anonymous closure> (dart:ui/hooks.dart:239:25)
E/flutter ( 7100): #11     _rootRun (dart:async/zone.dart:1126:13)
E/flutter ( 7100): #12     _CustomZone.run (dart:async/zone.dart:1023:19)
E/flutter ( 7100): #13     _runZoned (dart:async/zone.dart:1518:10)
E/flutter ( 7100): #14     runZoned (dart:async/zone.dart:1502:12)
E/flutter ( 7100): #15     _runMainZoned.<anonymous closure> (dart:ui/hooks.dart:231:5)
E/flutter ( 7100): #16     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:307:19)
E/flutter ( 7100): #17     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:174:12)
E/flutter ( 7100): 
E/flutter ( 7240): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: ServicesBinding.defaultBinaryMessenger was accessed before the binding was initialized.
E/flutter ( 7240): If you're running an application and need to access the binary messenger before `runApp()` has been called (for example, during plugin initialization), then you need to explicitly call the `WidgetsFlutterBinding.ensureInitialized()` first.
E/flutter ( 7240): If you're running a test, you can call the `TestWidgetsFlutterBinding.ensureInitialized()` as the first line in your test's `main()` method to initialize the binding.
E/flutter ( 7240): #0      defaultBinaryMessenger.<anonymous closure> (package:flutter/src/services/binary_messenger.dart:76:7)
E/flutter ( 7240): #1      defaultBinaryMessenger (package:flutter/src/services/binary_messenger.dart:89:4)
E/flutter ( 7240): #2      MethodChannel.binaryMessenger (package:flutter/src/services/platform_channel.dart:140:62)
E/flutter ( 7240): #3      MethodChannel.setMethodCallHandler (package:flutter/src/services/platform_channel.dart:368:5)
E/flutter ( 7240): #4      new Firestore (package:cloud_firestore/src/firestore.dart:13:13)
E/flutter ( 7240): #5      Firestore.instance (package:cloud_firestore/src/firestore.dart:39:37)
E/flutter ( 7240): #6      Firestore.instance (package:cloud_firestore/src/firestore.dart:39:26)
E/flutter ( 7240): #7      new UserRepository (package:example_application/repository/user.dart:11:49)
E/flutter ( 7240): #8      new ExampleApplicationApp (package:example_application/main.dart:33:42)
E/flutter ( 7240): #9      main (package:example_application/main.dart:29:10)
E/flutter ( 7240): #10     _runMainZoned.<anonymous closure>.<anonymous closure> (dart:ui/hooks.dart:239:25)
E/flutter ( 7240): #11     _rootRun (dart:async/zone.dart:1126:13)
E/flutter ( 7240): #12     _CustomZone.run (dart:async/zone.dart:1023:19)
E/flutter ( 7240): #13     _runZoned (dart:async/zone.dart:1518:10)
E/flutter ( 7240): #14     runZoned (dart:async/zone.dart:1502:12)
E/flutter ( 7240): #15     _runMainZoned.<anonymous closure> (dart:ui/hooks.dart:231:5)
E/flutter ( 7240): #16     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:307:19)
E/flutter ( 7240): #17     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:174:12)
E/flutter ( 7240): 
E/flutter ( 7576): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: ServicesBinding.defaultBinaryMessenger was accessed before the binding was initialized.
E/flutter ( 7576): If you're running an application and need to access the binary messenger before `runApp()` has been called (for example, during plugin initialization), then you need to explicitly call the `WidgetsFlutterBinding.ensureInitialized()` first.
E/flutter ( 7576): If you're running a test, you can call the `TestWidgetsFlutterBinding.ensureInitialized()` as the first line in your test's `main()` method to initialize the binding.
E/flutter ( 7576): #0      defaultBinaryMessenger.<anonymous closure> (package:flutter/src/services/binary_messenger.dart:76:7)
E/flutter ( 7576): #1      defaultBinaryMessenger (package:flutter/src/services/binary_messenger.dart:89:4)
E/flutter ( 7576): #2      MethodChannel.binaryMessenger (package:flutter/src/services/platform_channel.dart:140:62)
E/flutter ( 7576): #3      MethodChannel.setMethodCallHandler (package:flutter/src/services/platform_channel.dart:368:5)
E/flutter ( 7576): #4      new Firestore (package:cloud_firestore/src/firestore.dart:13:13)
E/flutter ( 7576): #5      Firestore.instance (package:cloud_firestore/src/firestore.dart:39:37)
E/flutter ( 7576): #6      Firestore.instance (package:cloud_firestore/src/firestore.dart:39:26)
E/flutter ( 7576): #7      new UserRepository (package:example_application/repository/user.dart:11:49)
E/flutter ( 7576): #8      new ExampleApplicationApp (package:example_application/main.dart:33:42)
E/flutter ( 7576): #9      main (package:example_application/main.dart:29:10)
E/flutter ( 7576): #10     _runMainZoned.<anonymous closure>.<anonymous closure> (dart:ui/hooks.dart:239:25)
E/flutter ( 7576): #11     _rootRun (dart:async/zone.dart:1126:13)
E/flutter ( 7576): #12     _CustomZone.run (dart:async/zone.dart:1023:19)
E/flutter ( 7576): #13     _runZoned (dart:async/zone.dart:1518:10)
E/flutter ( 7576): #14     runZoned (dart:async/zone.dart:1502:12)
E/flutter ( 7576): #15     _runMainZoned.<anonymous closure> (dart:ui/hooks.dart:231:5)
E/flutter ( 7576): #16     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:307:19)
E/flutter ( 7576): #17     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:174:12)
E/flutter ( 7576): 
E/flutter ( 7772): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: ServicesBinding.defaultBinaryMessenger was accessed before the binding was initialized.
E/flutter ( 7772): If you're running an application and need to access the binary messenger before `runApp()` has been called (for example, during plugin initialization), then you need to explicitly call the `WidgetsFlutterBinding.ensureInitialized()` first.
E/flutter ( 7772): If you're running a test, you can call the `TestWidgetsFlutterBinding.ensureInitialized()` as the first line in your test's `main()` method to initialize the binding.
E/flutter ( 7772): #0      defaultBinaryMessenger.<anonymous closure> (package:flutter/src/services/binary_messenger.dart:76:7)
E/flutter ( 7772): #1      defaultBinaryMessenger (package:flutter/src/services/binary_messenger.dart:89:4)
E/flutter ( 7772): #2      MethodChannel.binaryMessenger (package:flutter/src/services/platform_channel.dart:140:62)
E/flutter ( 7772): #3      MethodChannel.setMethodCallHandler (package:flutter/src/services/platform_channel.dart:368:5)
E/flutter ( 7772): #4      new Firestore (package:cloud_firestore/src/firestore.dart:13:13)
E/flutter ( 7772): #5      Firestore.instance (package:cloud_firestore/src/firestore.dart:39:37)
E/flutter ( 7772): #6      Firestore.instance (package:cloud_firestore/src/firestore.dart:39:26)
E/flutter ( 7772): #7      new UserRepository (package:example_application/repository/user.dart:11:49)
E/flutter ( 7772): #8      new ExampleApplicationApp (package:example_application/main.dart:33:42)
E/flutter ( 7772): #9      main (package:example_application/main.dart:29:10)
E/flutter ( 7772): #10     _runMainZoned.<anonymous closure>.<anonymous closure> (dart:ui/hooks.dart:239:25)
E/flutter ( 7772): #11     _rootRun (dart:async/zone.dart:1126:13)
E/flutter ( 7772): #12     _CustomZone.run (dart:async/zone.dart:1023:19)
E/flutter ( 7772): #13     _runZoned (dart:async/zone.dart:1518:10)
E/flutter ( 7772): #14     runZoned (dart:async/zone.dart:1502:12)
E/flutter ( 7772): #15     _runMainZoned.<anonymous closure> (dart:ui/hooks.dart:231:5)
E/flutter ( 7772): #16     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:307:19)
E/flutter ( 7772): #17     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:174:12)
E/flutter ( 7772): 
registerResGeneratingTask is deprecated, use registerGeneratedResFolders(FileCollection)
registerResGeneratingTask is deprecated, use registerGeneratedResFolders(FileCollection)
registerResGeneratingTask is deprecated, use registerGeneratedResFolders(FileCollection)
Application finished.`

**Comment by amirh** [comment]
> flutter pub update - doesn't exist ( error for console `Could not find a subcommand named "update" for "flutter pub".`)

Sorry I meant `flutter pub upgrade`

@akarabach

**Comment by blasten** [comment]
I was able to reproduce the issue described in #46807. It's caused by a combination of issues.

1. Flutter apps have a hard-coded `settings.gradle` that imports plugins as subprojects. The current script doesn't account for a plugin that doesn't have an `android` directory.
2. Android Studio may create an `android` directory in a subproject if it doesn't have one. This is due to 1.
3. If your app isn't using AndroidX. We started showing a warning that asks users to open Android Studio and migrate to AndroidX.
4. Android Studio migrates the app to AndroidX, but it also creates an empty `android` directory.
5.  f264d6c just skips plugins if they don't have an `android/` directory.
6.  now you get:
```
Could not determine the dependencies of task ':shared_preferences:compileReleaseAidl'.
> Could not resolve all task dependencies for configuration ':shared_preferences:releaseCompileClasspath'.
   > Could not resolve project :shared_preferences_macos.
     Required by:
         project :shared_preferences
      > Unable to find a matching configuration of project :shared_preferences_macos:
          - None of the consumable configurations have attributes.
```
cc @amirh

**Comment by akarabach** [comment]
> > flutter pub update - doesn't exist ( error for console `Could not find a subcommand named "update" for "flutter pub".`)
> 
> Sorry I meant `flutter pub upgrade`
> 
> @akarabach

Doesnt help to me ;(

**Comment by amirh** [comment]
@akarabach The `ServicesBinding.defaultBinaryMessenger was accessed before the binding was initialized` seems unrelated to the current issue (your app does build).

If you believe it's a flutter bug please file a separate issue.

FWIW there was a breaking change that required the widgets binding to be initialized before using method channels, so it may be that your application is using a plugin before invoking `runApp`.

**Comment by blasten** [comment]
As a consequence, users that run `flutter pub cache repair` or deleted the `.pub-cache` directory got the the plugins from pub  again without the `android/` directory. This is why that fixed the problem too.

**Comment by akarabach** [comment]
Wipe data on android emulator helped me, maybe it would help someone else

**Comment by alexdacoda** [comment]
my deadline is freaking tomorrow!!!!!!!! and then this crazy update comes up and breaks the app. why dont you people warn users before the update i am on the freaking stable channel for God sake.

**Comment by feinstein** [comment]
1 - You can't warn for something that you don't know exists.

2 - Just roll back to the previous Flutter version and you will be fine.

**Comment by alexdacoda** [comment]
how do i do that i may have already switched channel and done flutter upgrade over 10 times checking back and forth if all the channels have the same problem

**Comment by DaYuClo** [comment]
I just want to add that I didn't update my Flutter or anything last week but my app was broken.  I couldn't get it to run or build it for Android.  I tried again this morning without updating or roll back or do anything and it is working again. 

 Flutter (Channel dev, v1.12.13, on Mac OS X 10.15.1 19B88, locale en-US)
    • Flutter version 1.12.13 
    • Framework revision cf37c2cd07 (3 weeks ago), 2019-11-25 12:04:30 -0800
    • Engine revision b6b54fd606
    • Dart version 2.7.0

**Comment by feinstein** [comment]
@alexdacoda try this: https://stackoverflow.com/a/49468364/1227166

**Comment by alexdacoda** [comment]
Thanks its working now, sooner or later i will soon need to update, how to i know when all the issues have been fixed

**Comment by feinstein** [comment]
This issue would be marked as closed, not as open, so you try it again and if it doesn't work you roll back your version one more time and open an issue to let the Flutter team know.

**Comment by amirh** [comment]
@alexdacoda as far as I know the issue has been resolved. If you're still hitting these Gradle errors after running `flutter pub upgrade` could you please share the error log? (if you could also share the dependency section of your `pubspec.yaml` that should provide even more insight). Thanks.

**Comment by Icyoung** [comment]
@amirh  the flutter plugin connectivity has the error yet.

**Comment by amirh** [comment]
@Icyoung can you please share the error you're seeing, and the connectivity entry from your local pubspec.lock ?

**Comment by sd-timaru** [comment]
I ran `flutter pub upgrade` but i still have plugins that generate this error:


```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':devicelocale:verifyReleaseResources'.
> A failure occurred while executing com.android.build.gradle.internal.tasks.Workers$ActionFacade
   > Android resource linking failed
     C:\Users\Administrator\.gradle\caches\transforms-2\files-2.1\fb78fd57204adb81b79296ffb0fd6b51\core-1.0.0\res\values\values.xml:57:5-88:25: AAPT: error: resource android:attr/fontVariationSettings not found
     C:\Users\Administrator\.gradle\caches\transforms-2\files-2.1\fb78fd57204adb81b79296ffb0fd6b51\core-1.0.0\res\values\values.xml:57:5-88:25: AAPT: error: resource android:attr/ttcIndex not found.
```


`The plugin device_apps could not be built due to the issue above.
`
pubspec.lock:
`  device_apps:
    dependency: "direct main"
    description:
      name: device_apps
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.0.9"`

`The plugin firebase_messaging could not be built due to the issue above.
`
pubspec.lock:
`  firebase_messaging:
    dependency: "direct main"
    description:
      name: firebase_messaging
      url: "https://pub.dartlang.org"
    source: hosted
    version: "6.0.9"`

**Comment by amirh** [comment]
@sd-timaru this seems like a different issue, seems similar to: https://github.com/flutter/flutter/issues/46881

**Comment by amirh** [comment]
Postmortem is here: https://github.com/flutter/flutter/wiki/Postmortem:-Widespread-Gradle-failures-on-Flutter-v1.12.13-hotfix.5

**Comment by feinstein** [comment]
Excellent postmortem, I just finished it right now and I wonder if there should be an additional build test for the Flutter CI, where a simple app that includes the top 25 packages should build successfully. 

This would work as a safety net to ensure any flutter upgrades won't break the general workflow of the majority of apps.

Of course this would trigger on breaking changes, but could also serve as a double-checker for the breaking-changes docs, to see if all the build erros are documented there.

I just I am not sure if this will be really helpful as a "final line of defense" or if it will be a constant source of false alarms.

**Comment by amirh** [comment]
Thanks @feinstein .

Ensuring that Flutter changes are not breaking packages unintentionally is a great idea and will greatly improve the entire ecosystem. In fact that is the exact purpose of the [flutter/tests](https://github.com/flutter/tests) test suite. This is a test suite that runs against all Flutter changes, and any package author can add their package and make it run on Flutter's CI. It's new and currently includes a very few packages, but I hope and expect to see it grow.

Note that flutter/tests wouldn't have helped to mitigate the current issue, as CI currently does not use Android Studio to run the tests. Also, all `flutter/plugins` and `FirebaseExtended/FlutterFire` packages runs CI against Flutter master, so all the affected plugins for the current issue were already running on CI with the offending changes (lack of Android Studio on CI was the reason the issue wasn't detected).

**Comment by feinstein** [comment]
Perfect, thanks for the thorough explanation @amirh.
